File: src/main/java/com/hepdd/gtmthings/common/block/machine/multiblock/part/computation/WirelessOpticalComputationHatchMachine.java
package com.hepdd.gtmthings.common.block.machine.multiblock.part.computation;

import com.gregtechceu.gtceu.api.blockentity.MetaMachineBlockEntity;
import com.gregtechceu.gtceu.api.capability.IOpticalComputationHatch;
import com.gregtechceu.gtceu.api.machine.MetaMachine;
import com.gregtechceu.gtceu.api.machine.feature.IDataStickInteractable;
import com.gregtechceu.gtceu.api.machine.multiblock.part.WorkableMultiblockPartMachine;

import net.minecraft.MethodsReturnNonnullByDefault;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.BlockHitResult;

import com.hepdd.gtmthings.api.capability.IGTMTJadeIF;
import com.hepdd.gtmthings.common.block.machine.trait.WirelessNotifiableComputationContainer;
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted;
import lombok.Getter;

import javax.annotation.Nullable;
import javax.annotation.ParametersAreNonnullByDefault;

@Getter
@MethodsReturnNonnullByDefault
@ParametersAreNonnullByDefault
public class WirelessOpticalComputationHatchMachine extends WorkableMultiblockPartMachine implements IOpticalComputationHatch, IDataStickInteractable, IGTMTJadeIF {

    private final boolean transmitter;

    @Persisted
    @Nullable
    private BlockPos transmitterPos;
    @Persisted
    @Nullable
    private BlockPos receiverPos;
    protected WirelessNotifiableComputationContainer computationContainer;

    public WirelessOpticalComputationHatchMachine(MetaMachineBlockEntity holder, boolean transmitter) {
        super(holder);
        this.transmitter = transmitter;
        this.computationContainer = createComputationContainer(transmitter);
    }

    protected WirelessNotifiableComputationContainer createComputationContainer(Object... args) {
        if (args.length > 0 && args[args.length - 1] instanceof Boolean transmitter) {
            return new WirelessNotifiableComputationContainer(this, transmitter);
        }
        throw new IllegalArgumentException();
    }

    @Override
    public boolean shouldOpenUI(Player player, InteractionHand hand, BlockHitResult hit) {
        return false;
    }

    @Override
    public boolean canShared() {
        return false;
    }

    @Override
    public boolean isbinded() {
        return (this.transmitterPos != null || this.receiverPos != null);
    }

    @Override
    public String getBindPos() {
        if (this.isTransmitter() && this.receiverPos != null) {
            return this.receiverPos.toShortString();
        } else if (!this.isTransmitter() && this.transmitterPos != null) {
            return this.transmitterPos.toShortString();
        }
        return "";
    }

    @Override
    public long requestCWU(long cwu, boolean simulate) {
        return computationContainer.requestCWU(cwu, simulate);
    }

    @Override
    public long getMaxCWU() {
        return computationContainer.getMaxCWU();
    }

    @Override
    public boolean canBridge() {
        return computationContainer.canBridge();
    }

    private static final String KEY_TRANSMITTER = "wireless_computation_transmitter";
    private static final String KEY_RECEIVER = "wireless_computation_receiver";

    private void setTransmitterPos(BlockPos pos) {
        if (transmitterPos != null) {
            var level = getLevel();
            if (level != null) {
                if (MetaMachine.getMachine(level, transmitterPos) instanceof WirelessOpticalComputationHatchMachine machine) {
                    machine.receiverPos = null;
                }
            }
        }
        transmitterPos = pos;
    }

    private void setReceiverPos(BlockPos pos) {
        if (receiverPos != null) {
            var level = getLevel();
            if (level != null) {
                if (MetaMachine.getMachine(level, receiverPos) instanceof WirelessOpticalComputationHatchMachine machine) {
                    machine.transmitterPos = null;
                }
            }
        }
        receiverPos = pos;
    }

    private static CompoundTag createPos(BlockPos pos) {
        CompoundTag posTag = new CompoundTag();
        posTag.putInt("x", pos.getX());
        posTag.putInt("y", pos.getY());
        posTag.putInt("z", pos.getZ());
        return posTag;
    }

    private static BlockPos getPos(CompoundTag tag) {
        return new BlockPos(tag.getInt("x"), tag.getInt("y"), tag.getInt("z"));
    }

    @Override
    public InteractionResult onDataStickShiftUse(Player player, ItemStack dataStick) {
        if (isRemote()) return InteractionResult.SUCCESS;

        CompoundTag tag = dataStick.getOrCreateTag();
        BlockPos currentPos = getPos();
        if (isTransmitter()) {
            tag.put(KEY_TRANSMITTER, createPos(currentPos));
            player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_transmitter_hatch.tobind"));
        } else {
            tag.put(KEY_RECEIVER, createPos(currentPos));
            player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_receiver_hatch.tobind"));
        }
        dataStick.setHoverName(Component.translatable("gtceu.machine.me.import_part.data_stick.name", Component.translatable(this.getDefinition().getDescriptionId())));
        return InteractionResult.SUCCESS;
    }

    @Override
    public InteractionResult onDataStickUse(Player player, ItemStack dataStick) {
        if (isRemote()) return InteractionResult.sidedSuccess(true);

        CompoundTag tag = dataStick.getTag();
        if (tag == null) return InteractionResult.PASS;

        if (isTransmitter() && tag.contains(KEY_RECEIVER, 10)) {
            BlockPos otherPos = getPos(tag.getCompound(KEY_RECEIVER));
            if (bindWith(otherPos, player)) {
                return InteractionResult.SUCCESS;
            }
        } else if (!isTransmitter() && tag.contains(KEY_TRANSMITTER, 10)) {
            BlockPos otherPos = getPos(tag.getCompound(KEY_TRANSMITTER));
            if (bindWith(otherPos, player)) {
                return InteractionResult.SUCCESS;
            }
        }

        return InteractionResult.PASS;
    }

    private boolean bindWith(BlockPos otherPos, Player player) {
        Level level = getLevel();
        if (level == null || otherPos.equals(this.getPos())) return false;

        MetaMachine otherMachine = MetaMachine.getMachine(level, otherPos);
        if (otherMachine instanceof WirelessOpticalComputationHatchMachine otherWoch) {
            if (this.isTransmitter() == otherWoch.isTransmitter()) {
                return false;
            }
            if (isTransmitter()) {
                this.setReceiverPos(otherPos);
                otherWoch.setTransmitterPos(this.getPos());
            } else {
                this.setTransmitterPos(otherPos);
                otherWoch.setReceiverPos(this.getPos());
            }

            player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_hatch.binded"));
            return true;
        }
        return false;
    }
}

File: src/main/java/com/hepdd/gtmthings/common/block/machine/multiblock/part/computation/WirelessOpticalComputationHatchMachine.kt
package com.hepdd.gtmthings.common.block.machine.multiblock.part.computation

import com.gregtechceu.gtceu.api.capability.recipe.IO
import com.gregtechceu.gtceu.api.machine.IMachineBlockEntity
import com.gregtechceu.gtceu.api.machine.feature.IInteractedMachine
import com.gregtechceu.gtceu.api.machine.multiblock.part.MultiblockPartMachine
import com.gregtechceu.gtceu.common.data.GTItems
import com.hepdd.gtmthings.api.capability.IGTMTJadeIF
import com.hepdd.gtmthings.common.block.machine.trait.WirelessNotifiableComputationContainer
import com.lowdragmc.lowdraglib.syncdata.annotation.Persisted
import com.lowdragmc.lowdraglib.syncdata.field.ManagedFieldHolder
import net.minecraft.core.BlockPos
import net.minecraft.nbt.CompoundTag
import net.minecraft.network.chat.Component
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.player.Player
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.phys.BlockHitResult

open class WirelessOpticalComputationHatchMachine(holder: IMachineBlockEntity, transmitter: Boolean):MultiblockPartMachine(holder),IInteractedMachine, IGTMTJadeIF {

    companion object {
        @JvmStatic
        val MANAGED_FIELD_HOLDER: ManagedFieldHolder = ManagedFieldHolder(
            WirelessOpticalComputationHatchMachine::class.java, MultiblockPartMachine.MANAGED_FIELD_HOLDER
        )
    }


    override fun getFieldHolder(): ManagedFieldHolder {
        return MANAGED_FIELD_HOLDER
    }

    override fun isRemote(): Boolean {
        return level?.isClientSide ?: true
    }

    private var transmitter = false

    @Persisted
    var transmitterPos: BlockPos? = null

    @Persisted
    var receiverPos: BlockPos? = null
    var computationContainer: WirelessNotifiableComputationContainer? = null

    init {
        this.transmitter = transmitter
        this.computationContainer = createComputationContainer(transmitter)
    }


    protected fun createComputationContainer(vararg args: Any?): WirelessNotifiableComputationContainer {
        var io = IO.IN
        if (args.size > 1 && args[args.size - 2] is IO) {
            io = (args[args.size - 2] as IO)
        }
        if (args.isNotEmpty() && args[args.size - 1] is Boolean) {
            return WirelessNotifiableComputationContainer(this, io, transmitter)
        }
        throw IllegalArgumentException()
    }

    override fun shouldOpenUI(player: Player, hand: InteractionHand, hit: BlockHitResult): Boolean {
        return false
    }

    override fun canShared(): Boolean {
        return false
    }

    override fun onUse(
        state: BlockState,
        world: Level,
        pos: BlockPos,
        player: Player,
        hand: InteractionHand,
        hit: BlockHitResult
    ): InteractionResult {
        val `is` = player.getItemInHand(hand)
        if (`is`.isEmpty) return InteractionResult.PASS
        if (`is`.`is`(GTItems.TOOL_DATA_STICK.asItem())) {
            if (transmitter) {
                if (this.transmitterPos == null) this.transmitterPos = pos
                var tag = `is`.tag
                if (tag != null) {
                    val posTag = CompoundTag()
                    posTag.putInt("x", this.transmitterPos!!.x)
                    posTag.putInt("y", this.transmitterPos!!.y)
                    posTag.putInt("z", this.transmitterPos!!.z)
                    tag.put("transmitterPos", posTag)
                    val bindPos = tag.get("receiverPos") as CompoundTag?
                    if (bindPos != null) {
                        val recPos = BlockPos(bindPos.getInt("x"), bindPos.getInt("y"), bindPos.getInt("z"))
                        (getMachine(level!!,recPos) as? WirelessOpticalComputationHatchMachine)?.let { woc ->
                            if (!woc.transmitter) {
                                woc.transmitterPos = this.transmitterPos
                                this.receiverPos = recPos
                                tag.remove("transmitterPos")
                                tag.remove("receiverPos")
                                if (level!!.isClientSide()) {
                                    player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_hatch.binded"))
                                }
                            }
                        }
                    } else {
                        if (level!!.isClientSide()) {
                            player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_transmitter_hatch.tobind"))
                        }
                    }
                    `is`.tag = tag
                } else {
                    tag = CompoundTag()
                    val posTag = CompoundTag()
                    posTag.putInt("x", this.transmitterPos!!.x)
                    posTag.putInt("y", this.transmitterPos!!.y)
                    posTag.putInt("z", this.transmitterPos!!.z)
                    tag.put("transmitterPos", posTag)
                    `is`.tag = tag
                    if (level!!.isClientSide()) {
                        player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_transmitter_hatch.tobind"))
                    }
                }
                return InteractionResult.SUCCESS
            } else {
                if (this.receiverPos == null) this.receiverPos = pos
                var tag = `is`.tag
                if (tag != null) {
                    val posTag = CompoundTag()
                    posTag.putInt("x", this.receiverPos!!.x)
                    posTag.putInt("y", this.receiverPos!!.y)
                    posTag.putInt("z", this.receiverPos!!.z)
                    tag.put("receiverPos", posTag)
                    val bindPos = tag.get("transmitterPos") as CompoundTag?
                    if (bindPos != null) {
                        val tranPos = BlockPos(bindPos.getInt("x"), bindPos.getInt("y"), bindPos.getInt("z"))
                        (getMachine(level!!,tranPos) as? WirelessOpticalComputationHatchMachine)?.let { woc ->
                            if (!woc.transmitter) {
                                woc.receiverPos = this.receiverPos
                                this.transmitterPos = tranPos
                                tag.remove("transmitterPos")
                                tag.remove("receiverPos")
                                if (level!!.isClientSide()) {
                                    player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_hatch.binded"))
                                }
                            }
                        }
                    } else {
                        if (level!!.isClientSide()) {
                            player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_receiver_hatch.tobind"))
                        }
                    }
                    `is`.tag = tag
                } else {
                    tag = CompoundTag()
                    val posTag = CompoundTag()
                    posTag.putInt("x", this.receiverPos!!.x)
                    posTag.putInt("y", this.receiverPos!!.y)
                    posTag.putInt("z", this.receiverPos!!.z)
                    tag.put("receiverPos", posTag)
                    `is`.tag = tag
                    if (level!!.isClientSide()) {
                        player.sendSystemMessage(Component.translatable("gtmthings.machine.wireless_computation_receiver_hatch.tobind"))
                    }
                }
                return InteractionResult.SUCCESS
            }
        }
        return InteractionResult.PASS
    }

    override fun isTransmitter(): Boolean {
        return transmitter
    }

    override fun isbinded(): Boolean {
        return (this.transmitterPos != null || this.receiverPos != null)
    }

    override fun getBindPos(): String {
        if (this.isTransmitter() && this.receiverPos != null) {
            return this.receiverPos!!.toShortString()
        } else if (!this.isTransmitter() && this.transmitterPos != null) {
            return this.transmitterPos!!.toShortString()
        }
        return ""
    }
}